doc/Container Hardening/SONiC_container_hardening_HLD.md

```diff
@@ -0,0 +1,349 @@
+	Current: cap_chown,cap_dac_override,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_net_bind_service,cap_net_raw,cap_sys_chroot,cap_mknod,cap_audit_write,cap_setfcap=eip
+
+If, for some reason, a docker must retain a specific capablity functionality on top of the container (which is removed after removing the `--privileged` flag), we can do that with the following:
+
+In the docker-database.mk file adjust this line:
+
+        $(DOCKER_DATABASE)_RUN_OPT += -t –-cap-add NET_ADMIN  # Changed by removing the --privileged flag and adding --cap-add flag
+
+### 4.2 net=host optimization
+
+Here we will provide a detailed example of how to switch from the `--net=host` configuration (host network) to the `--net=bridge` configuration paired with port forwarding in a specific container. We are using the database container as an example for this item.
+
+The original docker creation should be like in the example below:
+
+docker with host sharing:
+
+	docker create --privileged -t -v /etc/sonic:/etc/sonic:ro  \
+					--net=$NET \
+					-e RUNTIME_OWNER=local \
+					--uts=host \
+					--log-opt max-size=2M --log-opt max-file=5 \
+					--tmpfs /tmp \
+					$DB_OPT \
+					$REDIS_MNT \
+					-v /usr/share/sonic/device/$PLATFORM:/usr/share/sonic/platform:ro \
+					--tmpfs /var/tmp \
+					--env "NAMESPACE_ID"="$DEV" \
+					--env "NAMESPACE_PREFIX"="$NAMESPACE_PREFIX" \
+					--env "NAMESPACE_COUNT"=$NUM_ASIC \
+					--name=database_no_net \
+					--cap-drop=NET_ADMIN \
+					docker-database:latest
+
+To disable the sharing of the networking stack between the host and a container we need to remove the flag: `--net=host`. Because we have not specified any `--network` flag, the containers connect to the default bridge network `--net=bridge`.
+To support port forwarding we are required to add the flag:  `-p <port>:<port>`
+
+The "new" docker creation file database.sh can be seen in the code block below:
+
+Docker with port forwarding and default bridge network
+
+	docker create --privileged -t -v /etc/sonic:/etc/sonic:ro  \
+					**-p 6379:6379** \
+					-e RUNTIME_OWNER=local \
+					--uts=host \
+					--log-opt max-size=2M --log-opt max-file=5 \
+					--tmpfs /tmp \
+					$DB_OPT \
+					$REDIS_MNT \
+					-v /usr/share/sonic/device/$PLATFORM:/usr/share/sonic/platform:ro \
+					--tmpfs /var/tmp \
+					--env "NAMESPACE_ID"="$DEV" \
+					--env "NAMESPACE_PREFIX"="$NAMESPACE_PREFIX" \
+					--env "NAMESPACE_COUNT"=$NUM_ASIC \
+					--name=$DOCKERNAME \
+					docker-database:latest \
+
+**How we did it?**
+
+To create a docker with the flags above it is required to set the "new" flag in the file docker_image_ctl.js. Follow the call `docker create {{docker_image_run_opt}} \`: 
+and replace the `–--net=$NET`.
+docker flag generation
+
+	{%- if docker_container_name != "database" %}
+					--net=$NET \
+	{%- endif %}
+	{%- if docker_container_name == "database" %}
+					-p 6379:6379 \
+	{%- endif %}
+
+#### How to check?
```

Comment:
What is the user experience with this change with the current container capabilities? What is the plan to mitigate this?

Reply:
User should not experience any changes.
If there are, then one of the following should occur:
1. Adjust the user action/container action so that it does not require the elevated privileges or specific net host capabilities.
2. Add to the container the required capabilities it needs to perform.

We should go with option 1 first, and only after we have decided option 1 is not feasible - option 2 is on the table.